src/adapter/src/explain_new/optimizer_trace.rs

```diff
@@ -0,0 +1,181 @@
+            )?,
+            self.drain_explainable_entries::<DataflowDescription<LirRelationExpr>>(
+                &format, &context
+            )?,
+        ));
+
+        // Drain trace entries of type String, HirScalarExpr, MirScalarExpr
+        // which are useful for ad-hoc debugging.
+        results.extend(itertools::chain!(
+            self.drain_scalar_entries::<HirScalarExpr>(),
+            self.drain_scalar_entries::<MirScalarExpr>(),
+            self.drain_string_entries(),
+        ));
+
+        // sort plans by instant (TODO: this can be implemented in a more
+        // efficient way, as we can assume that each of the runs that are used
+        // to `*.extend` the `results` vector is already sorted).
+        results.sort_by_key(|x| x.instant);
+
+        Ok(results)
+    }
+
+    /// Collect all trace entries of a plan type `T` that implements
+    /// [`Explainable`].
+    fn drain_explainable_entries<T>(
+        &self,
+        format: &ExplainFormat,
+        context: &ExplainContext,
+    ) -> Result<Vec<TraceEntry<String>>, ExplainError>
+    where
+        T: Debug + 'static,
+        for<'a> Explainable<'a, T>: Explain<'a, Context = ExplainContext<'a>>,
+    {
+        if let Some(trace) = self.0.downcast_ref::<PlanTrace<T>>() {
+            trace
+                .drain_as_vec()
+                .into_iter()
+                .map(|mut entry| {
+                    Ok(TraceEntry {
+                        instant: entry.instant,
+                        duration: entry.duration,
+                        path: entry.path,
+                        plan: Explainable::new(&mut entry.plan).explain(
+                            format,
+                            context.config,
+                            context,
+                        )?,
+                    })
+                })
+                .collect()
+        } else {
+            Ok(vec![])
+        }
+    }
+
+    /// Collect all trace entries of a plan type `T` that implements
+    /// [`Displayable`].
+    fn drain_scalar_entries<T>(&self) -> Vec<TraceEntry<String>>
+    where
+        T: Debug + 'static,
+        for<'a> Displayable<'a, T>: DisplayText<()>,
+    {
+        if let Some(trace) = self.0.downcast_ref::<PlanTrace<T>>() {
+            trace
+                .drain_as_vec()
+                .into_iter()
+                .map(|entry| TraceEntry {
+                    instant: entry.instant,
+                    duration: entry.duration,
+                    path: entry.path,
+                    plan: text_string(&Displayable::from(&entry.plan)),
+                })
+                .collect()
+        } else {
+            vec![]
+        }
+    }
+
+    /// Collect all trace entries with plans of type [`String`].
+    fn drain_string_entries(&self) -> Vec<TraceEntry<String>> {
+        if let Some(trace) = self.0.downcast_ref::<PlanTrace<String>>() {
+            trace.drain_as_vec()
+        } else {
+            vec![]
+        }
```

Comment:
in what case do we actually hit this case? should we just panic? If we are draining entries, then its a logic bug in our code somewhere to have not entered the `OptimizerTrace` dispatch right?

Also i think what you have done here is reasonable, to just downcast the `Dispatch` itself to collect all the trace info into the final output!

Reply:
Yes, we definitely can panic here because these are private methods.

Reply:
nice